### PR TITLE
VIMC-4737 Export responsibilities summary for a touchstone

### PR DESCRIPTION
**Pull request by mwoodbri**
Add new `/touchstones/{touchstone-id}/responsibilities/csv/` endpoint that returns a tabular representation of annotated responsibilities, combining information available in full from the existing `/touchstones/{touchstone-id}/responsibilities/` and `/touchstones/{touchstone-id}/responsibilities/comments/` endpoints

This is required for vimc/montagu-webapps#465

To-do:
- [x] Revert `src/models` to `master` after merge of vimc/montagu-webmodels#98

### CODE REVIEW

File: src/app/src/main/kotlin/org/vaccineimpact/api/app/logic/ResponsibilitiesLogic.kt [hunk 1]
```diff
@@ -36,4 +38,59 @@ class RepositoriesResponsibilitiesLogic(
...
+        val sets = responsibilitiesRepository.getResponsibilitiesForTouchstone(touchstoneVersionId)
+        return sets.flatMap { set ->
+            val responsibilitySetComment = comments.find { it.modellingGroupId == set.modellingGroupId}?.comment
+            set.responsibilities.map { responsibility ->
+                val currentEstimateSet = responsibility.currentEstimateSet
+                val responsibilityComment = comments.find { it.modellingGroupId == set.modellingGroupId }?.responsibilities?.find { it.scenarioId == responsibility.scenario.id }?.comment
+                val responsibilityInfo = burdenEstimateRepository.getResponsibilityInfo(set.modellingGroupId, touchstoneVersionId, responsibility.scenario.id)
+                val expectationMapping = expectationsRepository.getExpectationsForResponsibility(responsibilityInfo.id)
+                val validatedRowMap = if (currentEstimateSet != null) burdenEstimateRepository.validateEstimates(currentEstimateSet, expectationMapping.expectation.expectedRowLookup()) else null
+                ResponsibilitiesRow(
```
Comment (EmmaLRussell) [inline src/app/src/main/kotlin/org/vaccineimpact/api/app/logic/ResponsibilitiesLogic.kt:54]: I think a couple of comments would be useful here, for the less obvious data derivations here, particularly `current_estimate_set_missing_estimates` - I think this returns the count of all the years and ages where the validatedRowMap returns false (invalid)?

File: src/app/src/main/kotlin/org/vaccineimpact/api/app/logic/ResponsibilitiesLogic.kt [hunk 2]
```diff
@@ -36,4 +38,59 @@ class RepositoriesResponsibilitiesLogic(
...
+                        responsibilityComment?.comment,
+                        responsibilityComment?.addedOn,
+                        responsibilityComment?.addedBy
+                )
+            }
+        }
+    }
+}
+
+data class ResponsibilitiesRow(
```
Comment (EmmaLRussell) [inline src/app/src/main/kotlin/org/vaccineimpact/api/app/logic/ResponsibilitiesLogic.kt:78]: These sorts of data classes usually live in /models

File: src/app/src/test/kotlin/org/vaccineimpact/api/tests/logic/ResponsibilitiesLogicTests.kt [hunk 3]
```diff
@@ -72,12 +77,77 @@ class ResponsibilitiesLogicTests : MontaguTests()
...
+                                    )
+                            )
+                    ))
+            )
+        }
+        val expectationMapping = ExpectationMapping(CountryOutcomeExpectations(77, "", 2001..2002, 97..99, CohortRestriction(), listOf(Country("UK", "United Kingdom")), listOf()), listOf(""), "Cholera")
+        val expectationsRepo = mock<ExpectationsRepository> {
+            on { getExpectationsForResponsibility(42) } doReturn expectationMapping
+        }
+        val burdenEstimateRepo = mock<BurdenEstimateRepository> {
```
Comment (EmmaLRussell) [inline src/app/src/test/kotlin/org/vaccineimpact/api/tests/logic/ResponsibilitiesLogicTests.kt:122]: Again, maybe one or two comments in this test about where the expectations for the less obvious parts of the results are coming from e.g. the count of missing estimates.

File: src/blackboxTests/src/test/kotlin/org/vaccineimpact/api/blackboxTests/tests/ResponsibilityCommentTests.kt [hunk 4]
```diff
@@ -96,4 +105,49 @@ class ResponsibilityCommentTests : DatabaseTest() {
...
+        assertThat(response.headers["Content-Type"]).isEqualTo("text/csv")
+        assertThat(response.headers["Content-Disposition"])
+                .isEqualTo("attachment; filename=\"responsibilities_touchstone-1.csv\"")
+
+        val schema = CSVSchema("ResponsibilitySetsWithComments")
+        schema.validate(response.text)
+
+        val csv = StringReader(response.text)
+                .use { CSVReader(it).readAll() }
+        val dataRow = csv.drop(1).first().toList()
```
Comment (EmmaLRussell) [inline src/blackboxTests/src/test/kotlin/org/vaccineimpact/api/blackboxTests/tests/ResponsibilityCommentTests.kt:132]: Could check the header row a well.

**Review by EmmaLRussell — COMMENTED** [review]
This all looks good to me as far as I can tell - a fiddly job! The logic is pretty dense so i think one or two comments would be useful.

From the ticket: 

_scenario_description | This will typically not include the disease name. It does in the example however, due to the names of the dummy scenarios._

I thought scenario description would be something like `yf-campaign`, so with disease id? Or is the scenario description like those already shown on the Responsibilities page e.g. 'Default - Routine and Campaign'? In that case, yes, good call to include `scenario_disease`.

**Comment by mwoodbri** [comment]
> From the ticket:
> 
> _scenario_description | This will typically not include the disease name. It does in the example however, due to the names of the dummy scenarios._
> 
> I thought scenario description would be something like `yf-campaign`, so with disease id? Or is the scenario description like those already shown on the Responsibilities page e.g. 'Default - Routine and Campaign'? In that case, yes, good call to include `scenario_disease`.

Good question! This may help:

table | column | value (prod) | value (dev)
-- | -- | -- | --
scenario | scenario_description | yf-routine-default | yf-routine
scenario_description | description | Default - Routine | Yellow Fever, routine
disease | name | Yellow Fever | Yellow Fever

File: src/blackboxTests/src/test/kotlin/org/vaccineimpact/api/blackboxTests/tests/ResponsibilityCommentTests.kt (hunk 4 above)
Comment (mwoodbri) [reply src/blackboxTests/src/test/kotlin/org/vaccineimpact/api/blackboxTests/tests/ResponsibilityCommentTests.kt:132]: That's done by the `schema.validate(response.text)` (L128)

**Pull request merged**